File: 22. Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        i = 0
        output = 2*n
        B = [i for i in range(n)]
        p = ""
        for k in range(0, output):
            if k in B:
                p = p + "("
            else:
                p = p + ")"
        A = [p]
        i = n-1
        j = i
        while True:
            if B[j] != 2*j: # adjustment b
                B[j] += 1
            else:
                while j != 0:
                    j -= 1
                    if B[j] != 2*j:
                        B[j] += 1
                        for m in range(j+1, n):
                            B[m] = B[m-1] + 1
                        j = n-1
                        break
                if j <= 0:
                    return A
                
            p = "" # then print
            for k in range(0, output):
                if k in B:
                    p = p + "("
                else:
                    p = p + ")"
            A.append(p)

File: 22. Generate Parentheses.cpp
class Solution { //https://discuss.leetcode.com/topic/26309/my-0ms-c-solution-without-recursion-and-dfs
public:
    vector<string> generateParenthesis(int n) {
        vector<string> parenTable;
        if (n == 0)
            return parenTable;
        string s;
        int pl = n, pr = n;
        for (int i = 0; i < n; i++) s += "(";
        for (int i = 0; i < n; i++) s += ")";  //s = "(((...()...)))"
        while (s != ""){
            parenTable.push_back(s);
            while (s != "" && (s.back() == ')' || pl - 1 <  pr + 1)){
                s.back() == ')' ? pr-- : pl --;
                s.pop_back();
            }
            if (s != ""){
                s.back() = ')';
                pl--, pr++;
                while (pl < n){
                    s += '(';
                    pl++;
                }
                while (pr < n){
                    s += ')';
                    pr++;
                }
            }
        }
        return parenTable;
    }
};
